students/rod_musser/lesson09/mailroom_oo/donor_models.py

```diff
@@ -0,0 +1,206 @@
+
+    def __eq__(self, other):
+        return self.name.lower() == other.name.lower()
+
+    def __lt__(self, other):
+        return self.sum_of_donations < other.sum_of_donations
+
+    @property
+    def name(self):
+        return self._name
+
+    @name.setter
+    def name(self, full_name):
+        self._name = full_name
+
+    @property
+    def num_of_donations(self):
+        """
+        Returns the number of donations this donor has made.
+        """
+        return len(self._donations)
+
+    @property
+    def sum_of_donations(self):
+        """
+        Returns the total amount of donations this donor has made
+        """
+        return sum(self._donations[:])
+
+    @property
+    def average_donation(self):
+        """
+        Returns the average number of donations this donor has made
+        """
+        return round(self.sum_of_donations / self.num_of_donations, 2)
+
+    @property
+    def last_donation(self):
+        """
+        Reutrns the most recent donation made by the donor
+        """
+        return self._donations[-1]
+
+    def add_donation(self, amount):
+        """
+        Add a donation to this donor's donation history
+
+        :param amount: The amount of the donation
+        """
+        self._donations.append(Donation(amount))
+
+    def create_thank_you_letter(self):
+        """
+        Returns a thank you letter for the donor's most recent donation.
+        """
+        return Letter(self).generate_letter()
+
+
+class DonorCollection:
+    """
+    Represents all the donors to the charity.  Provides methods to add a donor, search for a
+    donor, etc.
+    """
+    def __init__(self):
+        self._donors = {}
+
+    def add_donation(self, full_name, amount):
+        """
+        Adds a donation for the donor.  If the donor does not exist, it will create
+        a new Donor object.
+
+        :param full_name: The first and last name of the donor
+
+        :param amount: The amount of the donation.  It must be a positive number
+        """
+        self._donors.setdefault(full_name, Donor(full_name)).add_donation(amount)
+
+    def get_donor(self, full_name):
+        """
+        Returns donor for name
+
+        :param full_name: The first and last name of the donor
+        """
+        try:
+            return self._donors[full_name]
+        except KeyError:
+            return None
+
+    def has_donor(self, full_name):
+        """
+        Does case insensitve serach for donor name.
+        If found, returns correctly cased name.  Otherwise, returns None
+
+        :param full_name: The first and last name of the donor
+        """
+        for name in self._donors.keys():
+            if name.lower() == full_name.lower():
+                return name
+        return None
+
+    def list_donors(self):
+        """
+        Return a list of donor names with each name on a new line.
+        """
+        newLine = '\n'
+        return f"{newLine.join(self._donors)}"
```

Comment:
no need to use an f-string here, you can just do `return "\n".join(self._donors)`